turbo.json

```diff
@@ -44,6 +48,12 @@
       ],
       "outputs": [".next/**", "!.next/cache/**"]
     },
+    "database-design#build": {
+      "dependsOn": ["^build"],
+      "env": ["NEXT_PUBLIC_SUPABASE_URL", "NEXT_PUBLIC_SUPABASE_ANON_KEY", "OPENAI_API_KEY"],
```

Comment:
Do we need to care about these?

database-design
- [`VERCEL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/database-new/middleware.ts#L8)
- [`VERCEL_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/database-new/app/layout.tsx#L10-L11)

docs
- [`ANALYZE`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/docs/next.config.mjs#L13)
- [`NEXT_PUBLIC_BASE_PATH`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/docs/next.config.mjs#L41)
- [`NEXT_PUBLIC_DEV_AUTH_PAGE`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/docs/components/Navigation/NavigationMenu/TopNavBar.tsx#L85)

www
- [`ANALYZE`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/next.config.mjs#L36)
- [`NEXT_PUBLIC_LWX_GAME_WORD`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/components/LaunchWeek/X/Ticket/LWXGame.tsx#L22)
- [`NEXT_PUBLIC_VERCEL_ENV`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/lib/constants.ts#L4C36-L4C58)
- [`NEXT_PUBLIC_VERCEL_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/lib/constants.ts#L34)
- [`NEXT_PUBLIC_API_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/lib/constants.ts#L6C36-L6C55)
- [`NEXT_PUBLIC_STUDIO_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/lib/rewrites.js#L8C33-L8C55)
- [`NEXT_PUBLIC_DOCS_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/lib/rewrites.js#L16C33-L16C53)
- [`NEXT_PUBLIC_REFERENCE_DOCS_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/lib/rewrites.js#L30C33-L30C63)
- [`GITHUB_CHANGELOG_APP_REST_KEY`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/pages/changelog.tsx#L61C23-L61C52)
- [`GITHUB_CHANGELOG_APP_ID`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/pages/changelog.tsx#L104C28-L104C51)
- [`GITHUB_CHANGELOG_APP_INSTALLATION_ID`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/pages/changelog.tsx#L105C37-L105C73)
- [`GITHUB_CHANGELOG_APP_PRIVATE_KEY`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/pages/changelog.tsx#L106C33-L106C65)
- [`OPENAI_API_KEY`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/pages/api/ai/docs.ts#L8)
- [`NEXT_PUBLIC_SUPABASE_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/pages/api/ai/docs.ts#L9C33-L9C57)
- [`NEXT_PUBLIC_SUPABASE_ANON_KEY`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/www/pages/api/ai/docs.ts#L10C40-L10C69)
- [`NEXT_PUBLIC_HCAPTCHA_SITE_KEY`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/new/index.tsx#L86)

studio
- [`NEXT_PUBLIC_ENVIRONMENT`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/new/%5Bslug%5D.tsx#L102C41-L102C64)
- [`NEXT_PUBLIC_SUPABASE_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/api/ai/docs.ts#L9)
- [`NEXT_PUBLIC_SUPABASE_ANON_KEY`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/api/ai/docs.ts#L10)
- [`LOGFLARE_API_KEY`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/api/projects/%5Bref%5D/analytics/endpoints/%5Bname%5D.ts#L14)
- [`LOGFLARE_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/api/projects/%5Bref%5D/analytics/endpoints/%5Bname%5D.ts#L15C21-L15C33)
- [`DEFAULT_ORGANIZATION_NAME`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/api/profile/index.ts#L31C27-L31C52)
- [`AUTH_JWT_SECRET`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/api/projects/%5Bref%5D/config.ts#L27C19-L27C34)
- [`SUPABASE_URL`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/api/auth/%5Bref%5D/users/%5Bid%5D/factors.ts#L5C70-L5C90)
- [`SUPABASE_SERVICE_KEY`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/api/auth/%5Bref%5D/users/%5Bid%5D/factors.ts#L5C70-L5C90)
- [`OPENAI_API_KEY`](https://github.com/supabase/supabase/blob/a348cf0d0acb3bbb50d93a47c24cb3987f1799a8/apps/studio/pages/api/ai/sql/debug.ts#L6)

(might be more I missed)

Reply:
You're right, I thought turbo strict mode will catch them all but they'll just be undefined in the build. `OPEN_API_KEY` in `database-design` and `SUPABASE_KEY` in `www` threw an error because they're used during the build. We'll have to declare all env vars before merging this PR. I'll add another PR with an [eslint rule](https://turbo.build/repo/docs/reference/eslint-config-turbo) now.

Reply:
I just figured out that we don't need to specify all env vars because only the `NEXT_PUBLIC_*` vars are inlined in the build. The others are still left to be sourced from the [environment during runtime](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#runtime-environment-variables). The [framework inference](https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables#framework-inference) does most of the work.

`OPEN_API_KEY` in `database-design` and `SUPABASE_KEY` in `www` are special cases because they're used to fetch data during build time.

In short, we should remove almost all env vars from `turbo.json`.

Reply:
I deep-dived into this issue and the solution is more nuanced - only `NEXT_PUBLIC_*` are inlined in the bundle so they should should be declared as cache keys (`env` key in a task). The other env vars are available at runtime (when the next app is running). 

BUT, we do use some env vars in the `next.config.js` and those should also be cache keys because they change the build artifact.

For now, I've listed all env vars as cache keys, just to be on the safe side. We can move some of them into `passthroughEnv` when we're sure everything works.

Reply:
Makes sense - nice that they support wildcard env vars. Looks good 👍